### ISSUE TITLE
Circular reference issue

### ISSUE THREAD

**Issue by nisbus** [tomas-abrahamsson/gpb#194]
I have a situation where gpb is detecting the same message twice due to circular references.  
I managed to create a small sample to reproduce although it's not exactly the same as the one I'm encountering.  

directory structure:
```
   proto/
       v0/
           test.proto
           nest/
                nest.proto
```  
**test.proto**
```
syntax = "proto3";

import "proto/v0/nest/nest.proto";

message TestMessage {
  uint32 id = 1;
  string description = 2;
  NestedMessage nested = 3;
}
```

**nest.proto**
```
syntax = "proto3";
import "proto/v0/test.proto";

message NestedMessage {
  TestMessage test = 1;  
}
```  
  
When I try to compile those I get the following error:
```
gpb/bin/protoc-erl -I `pwd`/proto/v0/ -o-erl src -o-hrl include -pkgs -json `pwd`/proto/v0/test.proto

message name TestMessage defined more than once
```
  
Other generators don't have this issue as far as I can tell (python/c++).  
  
Thanks

**Comment by tomas-abrahamsson** [comment]
I will take a look

**Comment by tomas-abrahamsson** [comment]
For a quick workaround: The error goes away if called as
```
gpb/bin/protoc-erl -I `pwd`/proto/v0/ -o-erl src -o-hrl include -pkgs -json proto/v0/test.proto
instead of
gpb/bin/protoc-erl -I `pwd`/proto/v0/ -o-erl src -o-hrl include -pkgs -json `pwd`/proto/v0/test.proto
 ie omitting this ----------------------------------------------------------^^^^^^
```
This is because the function that checks for imports already seen, does not realize that
`/full/path/to/proto/v0/test.proto` (from the command line) is in fact the same file as `proto/v0/test.proto` (from nest's import)

I consider this a bug and will look into fixing it.

**Comment by tomas-abrahamsson** [comment]
It's been a long time, I thought I should keep you posted on progress: this one has prompted me to rethink and redesign some import processing in gpb, and also to investigate details on how this is handled in protoc. I'm still in the process of reworking things, getting all use-cases right, so it will be a little more before I have something working, hope that's ok?

**Comment by nisbus** [comment]
thanks for the update.
I've managed a workaround by removing dependencies from my files (using sed ;))
it's great of this makes gpb better in the end.
Let me know if I can help in any way.
At least I can run the new implementation against my files which are probably complex enough to find any issues.

**Comment by nisbus** [comment]
Hi Tomas,

I can verify that the solution is working for my code at least (I haven't tested it but generating them is done without errors, I assume it will just work).

Thanks for your great work, I'll keep in mind the limitations for symlinks and it might be a good idea to add info on that to the README (at least until you find a way to discern some method of detecting those)

Thanks again
nisbus

**Comment by tomas-abrahamsson** [comment]
Thanks for the quick testing! I'll include it in a release soon. Better symlink handling/dereferencing, will have to wait until some later release.

**Comment by tomas-abrahamsson** [comment]
Hi again, this is now included in 4.16.2, which was just released.
